CE-865 - Update the Consul Autopilot doc page (#1427)
hashicorp/web-unified-docs: content/consul/v1.21.x/content/docs/manage/scale/autopilot.mdx

---
layout: docs
page_title: Consul Autopilot
description: >-
  Use Autopilot features to monitor the Raft cluster, introduce stable servers, and clean up dead servers.
---

# Consul Autopilot

> This page describes Consul Autopilot, which supports automatic, operator-friendly management of Consul servers. Autopilot helps maintain the health and stability of the Consul server cluster by monitoring server health, introducing stable servers, and cleaning up dead servers. Furthermore, Consul Enterprise customers can leverage two additional Autopilot features, namely redundancy zones and automated upgrades to enhance datacenter resiliency and simplify operations.

Reviewer: ```suggestion
This page describes Consul Autopilot, a set of features that provide operator-friendly management automations for Consul servers. 
```

## Overview

> Autopilot includes the following features:

Reviewer: ```suggestion
Consul autopilot helps you maintain the health and stability of the Consul server cluster. It includes the following features:
```
Moved the sentence so that there's less at the top and more before the list.

</Tab>
</Tabs>

Changes to the Autopilot configuration are persisted in the Raft database maintained by the Consul servers. This means that Autopilott configuration will be included in the Consul snapshot data.

## Workflow

> ### Server health checking

Reviewer: ```suggestion
## Server health checking
```

An internal health check runs on the leader to track the stability of servers. A server is considered healthy if all of the following conditions are true.

- It has a SerfHealth status of 'Alive'.

> - The time since its last contact with the current leader is below `LastContactThreshold` (that by default is `200ms`).

Reviewer: ```suggestion
- The time since its last contact with the current leader is below `LastContactThreshold`. The default value is `200ms`.
```

- Its latest Raft term matches the leader's term.
- The number of Raft log entries it trails the leader by does not exceed `MaxTrailingLogs` (that by default is `250`).

> The status of these health checks can be viewed through the `/v1/operator/autopilot/health` HTTP endpoint, with a top level `Healthy` field indicating the overall status of the datacenter:

Reviewer: ```suggestion
To return the status of these health checks, use the `/v1/operator/autopilot/health` HTTP endpoint. The `Healthy` field at the top indicates the overall status of the datacenter:
```

| `LastContactThreshold`    | Duration | `200ms`       | Time duration that defines the maximum time since the last contact with the current leader for a server to be considered healthy. |
| `MaxTrailingLogs`         | Integer  | `250`         | Maximum number of Raft log entries that a server can trail the leader by and still be considered healthy.                         |
| `MinQuorum`               | Integer  | `0`           | Minimum number of healthy voting servers required to maintain quorum in the datacenter.                                           |
| `ServerStabilizationTime` | Duration | `10s`         | Time duration that a new server must remain healthy before it can become a voting member.                                         |
| `RedundancyZoneTag`       | String   | `""`          | Tag name used to identify redundancy zones for servers in Consul Enterprise.                                                      |
| `DisableUpgradeMigration` | Boolean  | `false`       | Flag to disable automatic upgrade migrations in Consul Enterprise.                                                                |

> | `UpgradeVersionTag`       | String   | `""`          | Tag name used to identify server versions for automated upgrades in Consul Enterprise.                                            |

Reviewer: ```suggestion
The following table lists autopilot configuration parameters, their descriptions, and their default values:

| Autopilot setting         | Type     | Default value | Description                                                                                                                       | 
| :------------------------ | :------- | :------------ | :-------------------------------------------------------------------------------------------------------------------------------- | 
| `CleanupDeadServers`      | Boolean  | `true`        | Enables periodic dead server removal from the Raft peer set.                                                              |
| `LastContactThreshold`    | Duration | `200ms`       | The interval that can elapse between a server's last contact with the current leader before Consul considers it unhealthy. |
| `MaxTrailingLogs`         | Integer  | `250`         | Maximum number of Raft log entries that a server can trail the leader by and still be considered healthy.                         |
| `MinQuorum`               | Integer  | `0`           | Minimum number of healthy voting servers required to maintain quorum in the datacenter.                                           |
| `ServerStabilizationTime` | Duration | `10s`         | Time duration that a new server must remain healthy before it can become a voting member.                                         |
| `RedundancyZoneTag`       | String   | `""`          | Tag name used to identify redundancy zones for servers in Consul Enterprise.                                                      |
| `DisableUpgradeMigration` | Boolean  | `false`       | Flag to disable automatic upgrade migrations in Consul Enterprise.                                                                |
| `UpgradeVersionTag`       | String   | `""`          | Tag name used to identify server versions for automated upgrades in Consul Enterprise.                                            |
```

## Server stabilization time

When a new server is added to the datacenter, there is a waiting period where it must be healthy and stable for a certain amount of time before being promoted to a full, voting member. This is defined by the `ServerStabilizationTime`, Autopilot's parameter, and by default is 10 seconds.

> In case your configuration requires a different amount of time, you can tune the parameter and assign it a different duration.

Reviewer: ```suggestion
If you need a different amount of time, you can tune the parameter to set a different duration. The following example extends the waiting period to 15 seconds:
```

UpgradeVersionTag = ""
```

### Dead server cleanup

> If Autopilot is disabled, it will take 72 hours for dead servers to be automatically reaped, or an operator must manually issue the `consul force-leave <dead-server-name>` command. If another server failure occurred it could jeopardize the quorum, even if the failed Consul server had been automatically replaced. Autopilot helps prevent these kinds of outages by quickly removing failed servers as soon as a replacement Consul server comes online. When servers are removed by the cleanup process they will enter the "left" state.

Reviewer: ```suggestion
When autopilot is disabled, it takes 72 hours for Consul to automatically reap dead servers. The alternative would be for an operator to manually issue the `consul force-leave <dead-server-name>` command for each dead server. 

In this situation, another server failure could jeopardize the cluster's quorum. The Consul cluster still considers the missing server a member of the datacenter, even if the failed Consul server was automatically replaced. 

Autopilot helps prevent these kinds of outages from becoming outages. It quickly removes failed servers as soon as a replacement Consul server comes online. When servers are removed by the cleanup process, they enter the "left" state and are not considered for the datacenter's quorum.
```

> With Autopilot's dead server cleanup enabled, dead servers will periodically be cleaned up and removed from the Raft peer set to prevent them from interfering with the quorum size and leader elections. The cleanup process will also be automatically triggered whenever a new server is successfully added to the datacenter.

Reviewer: ```suggestion
Autopilot also triggers the cleanup process automatically whenever a new server successfully joins the datacenter.
```

We suggest leaving the feature enabled to avoid faulty nodes remaining in the Raft pool for too long without the need for manual pruning. In test scenarios or in environments you can disable the faulty node pruning by using the `consul operator autopilot set-config -cleanup-dead-servers=false` command.

## Enterprise features

To further strengthen and automate Consul operations, there are two more Autopilot features in Consul Enterprise that customers can take advantage of to improve their datacenter resiliency and to simplify operations.

> ### Redundancy zones

Reviewer: ```suggestion
## Redundancy zones (Enterprise)
```

Consul’s redundancy zones provide high availability in the case of server failure through the Enterprise feature of Autopilot. Autopilot allows you to add read replicas to your datacenter that will be promoted to the "voting" status in case of voting server failure.

> You can utilise redundancy zones to implement isolated failure domains such as AWS Availability Zones (AZ), and therefore to obtain redundancy within an AZ without having to sustain the overhead of a large quorum.

Reviewer: ```suggestion
You can set up redundancy zones to implement isolated failure domains. For example, deploying a server and a read replica in each AWS Availability Zones (AZ) provides additional protection against failure within a region.
```

> Check [provide fault tolerance with redundancy zones](/consul/docs/manage/scale/redundancy-zone) to learn more on the functionality.

Reviewer: ```suggestion
To learn more, refer to [provide fault tolerance with redundancy zones](/consul/docs/manage/scale/redundancy-zone).
```

### Automated upgrades

Consul’s automatic upgrades provide a simplified way to upgrade existing Consul datacenter. This functionally is provided through the Enterprise feature of Autopilot. Autopilot allows you to add new servers directly to the datacenter and waits until you have enough servers running the new version to perform a leadership change and demote the old servers as "non-voters".

> Check [automate upgrades with Consul Enterprise](/consul/docs/upgrade/automated) to learn more on the functionality.

Reviewer: ```suggestion
To learn more, refer to [automate upgrades with Consul Enterprise](/consul/docs/upgrade/automated).
```
